File: BEmu/ReferenceDataRequest/RequestReferenceElementStringArray.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/RequestReferenceElementStringArray.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class RequestReferenceElementStringArray : Element
    {
        private readonly string _elementName;
        private readonly List<string> _values;

        internal RequestReferenceElementStringArray(string elementName)
        {
            this._elementName = elementName;
            this._values = new List<string>();
        }

        public override int NumElements { get { return 0; } }
        public override int NumValues { get { return this._values.Count; } }
        public override Name Name { get { return new Name(this._elementName); } }
        public override bool IsArray { get { return true; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsNull { get { return false; } }

        internal void AddValue(string value)
        {
            this._values.Add(value);
        }

        internal List<string> Values { get { return this._values; } }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1}[] = {{{2}", tabs, this.Name, Environment.NewLine);

            string secs = string.Join(", ", this._values);
            result.AppendFormat("{0}{1}{2}{3}", tabs, Types.IndentType.TAB, secs, Environment.NewLine);

            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }
    }
}

File: BEmu_cpp/src/ReferenceDataRequest/RequestReferenceElementStringArray.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/ReferenceDataRequest/RequestReferenceElementStringArray.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "ReferenceDataRequest/RequestReferenceElementStringArray.h"

#include "Types/IndentType.h"
#include "BloombergTypes/Name.h"

namespace BEmu
{
	namespace ReferenceDataRequest
	{
		RequestReferenceElementStringArray::RequestReferenceElementStringArray(const std::string& elementName)
		{
			this->_elementName = elementName;
		}

		RequestReferenceElementStringArray::~RequestReferenceElementStringArray()
		{
		}

		Name RequestReferenceElementStringArray::name() const { return Name(this->_elementName.c_str()); }
		size_t RequestReferenceElementStringArray::numValues() const { return this->_values.size(); }
		size_t RequestReferenceElementStringArray::numElements() const { return 0; }

		bool RequestReferenceElementStringArray::isNull() const { return false; }
		bool RequestReferenceElementStringArray::isArray() const { return true; }
		bool RequestReferenceElementStringArray::isComplexType() const { return false; }

		void RequestReferenceElementStringArray::addValue(const std::string& elementName)
		{
			this->_values.push_back(elementName);
		}

		std::vector<std::string> RequestReferenceElementStringArray::getValues() const
		{
			return this->_values;
		}

		std::ostream& RequestReferenceElementStringArray::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << this->_elementName << "[] = {" << std::endl;

			stream << tabs;
			for(size_t i = 0; i < this->_values.size(); i++)
			{
				stream << this->_values[i];

				if(i + 1 < this->_values.size())
					stream << ", ";
			}
			stream << std::endl;

			stream << tabs << '}' << std::endl;

			return stream;

		}

	}
}